src/test/java/org/terracotta/statistics/derived/latency/LatencyHistogramSimulator.java

```diff
@@ -0,0 +1,286 @@
+import org.jfree.chart.plot.XYPlot;
+import org.jfree.chart.renderer.xy.XYStepRenderer;
+import org.jfree.data.DomainOrder;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TestName;
+import org.junit.runner.RunWith;
+import org.terracotta.statistics.LongSample;
+import org.terracotta.statistics.Time;
+import org.terracotta.statistics.util.ConcurrentParameterized;
+import org.terracotta.statistics.util.ConcurrentParameterized.ConcurrencyLevel;
+import org.terracotta.statistics.util.GraphUtils;
+import org.terracotta.statistics.util.LongSampleDataset;
+import org.terracotta.statistics.util.Percentile;
+
+import java.awt.*;
+import java.io.File;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.stream.Collectors;
+import java.util.stream.DoubleStream;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import static java.lang.Math.round;
+import static java.lang.String.format;
+import static java.lang.System.lineSeparator;
+import static java.time.Duration.ofMillis;
+import static java.time.Duration.ofMinutes;
+import static java.time.Duration.ofSeconds;
+import static java.util.function.DoubleUnaryOperator.identity;
+import static java.util.stream.Collectors.toMap;
+import static java.util.stream.LongStream.iterate;
+import static org.junit.runners.Parameterized.Parameters;
+import static org.terracotta.statistics.LongSample.sample;
+import static org.terracotta.statistics.util.GraphUtils.writeCSV;
+import static org.terracotta.statistics.util.GraphUtils.writeGraph;
+import static org.terracotta.statistics.util.LatencyUtils.generateLatencies;
+import static org.terracotta.statistics.util.LatencyUtils.getPercentiles;
+import static org.terracotta.statistics.util.LatencyUtils.nsToMicros;
+import static org.terracotta.statistics.util.LatencyUtils.nsToSecs;
+
+/**
+ * Generates a PNG output from simulated latencies for different kind of histogram settings.
+ * <p>
+ * Goal is to simulate, for several kind of histogram configurations:
+ * <ol>
+ * <li>0-9 operations / second over 90 minutes (﻿5400 seconds)</li>
+ * <li>Each operation latencies are sent to the histogram implementation</li>
+ * <li>A thread is collecting the statistics each 10 seconds (min, max, 50%-ile, 95%-ile, 99%ile)</li>
+ * <li>This statistics are displayed in 3 graphs for the last 1-hour, last 30min and last 10min</li>
+ * <li>Each graph contains only 20-30 points (so for 10min, 1 point will aggregate 30sec of data and for 1-hour 180sec)</li>
+ * </ol>
+ *
+ * @author Mathieu Carbou
+ */
+@RunWith(ConcurrentParameterized.class)
+//@RunWith(Parameterized.class)
+@ConcurrencyLevel(2.0f)
+public class LatencyHistogramSimulator {
```

Comment:
Thought experiment... if I had written the derived statistics layer such that you didn't need to make any changes in this repository where would you have put these classes?

Reply:
the only possible place is tc-platform, because it is the only "common" place, but currently in tc-platform there's no such module that could accept this code because there's no stat stuff in there. Only management stuff.

Reply:
The thing is, this test class that doesn't really test anything extra is actually accounting for the bulk of the code in the PR...